src/join.rs

```diff
@@ -54,141 +54,48 @@ bitset_and!{A, B, C, D, E, F, G, H, I, J, K, L, M, N}
-    type Mask = &'a BitSet;
-    fn open(self) -> (Self::Mask, Self::Value) {
-        let (l, r) = (**self).open();
-        (l, GetRef(r))
-    }
+/// Join is an Iterator over a group of `Storages`
+pub struct Join<O: Open> {
+    keys: BitIter<O::Mask>,
+    values: O::Value,
 }
 
-impl<'b, 'a:'b, T> Open for &'b mut std::sync::RwLockWriteGuard<'a, T>
-    where T: Storage
-{
-    type Value = GetMut<'b, T::UnprotectedStorage>;
-    type Mask = &'b BitSet;
-    fn open(self) -> (Self::Mask, Self::Value) {
-        let (l, r) = (**self).open_mut();
-        (l, GetMut(r))
-    }
-}
-
-
-
-/// Get like `Open` provides the same feature of providing
-/// mutable vs immutable preserving around the UnprotectedStorage
-/// trait
-pub trait Get {
-    type Value;
-    unsafe fn get(&self, idx: Index) -> Self::Value;
-}
-
-pub struct GetRef<'a, T: 'a>(&'a T);
-impl<'a, T> Get for GetRef<'a, T>
-    where T: UnprotectedStorage
-{
-    type Value = &'a T::Component;
-    unsafe fn get(&self, idx: Index) -> Self::Value {
-        self.0.get(idx)
-    }
-}
-
-pub struct GetMut<'a, T: 'a>(&'a mut T);
-impl<'a, T> Get for GetMut<'a, T>
-    where T: UnprotectedStorage
-{
-    type Value = &'a mut T::Component;
-    #[allow(mutable_transmutes)]
-    unsafe fn get(&self, idx: Index) -> Self::Value {
-        // This is obviously unsafe and is one of the reasons this
-        // trait is marked as unsafe to being with. It is safe
-        // an an external api point of view because the bitmask
-        // iterator never visits the same index twice, otherwise
-        // this would provide multiple aliased mutable pointers which
-        // is illegal in rust.
-        let x: &mut Self = std::mem::transmute(self);
-        x.0.get_mut(idx)
-    }
-}
-
-/// Joined is an Iterator over a group of `Storages`
-pub struct Joined<K, V> {
-    keys: bitset::Iter<K>,
-    values: V,
-}
-
-impl<K, V> Joined<K, V>
-    where K: BitSetLike
-{
-    fn new(k: K, v: V) -> Joined<K, V> {
-        Joined {
-            keys: k.iter(),
-            values: v,
+impl<O: Open> From<O> for Join<O> {
+    fn from(o: O) -> Self {
+        let (keys, values) = o.open();
+        Join {
+            keys: keys.iter(),
+            values: values,
         }
     }
 }
 
-impl<K, V> std::iter::Iterator for Joined<K, V>
-    where K: BitSetLike,
-          V: Get
-{
-    type Item = V::Value;
-    fn next(&mut self) -> Option<Self::Item> {
-        if let Some(idx) = self.keys.next() {
-            unsafe { Some(self.values.get(idx)) }
-        } else {
-            None
-        }
+impl<O: Open> std::iter::Iterator for Join<O> {
+    type Item = O::Type;
+    fn next(&mut self) -> Option<O::Type> {
+        self.keys.next().map(|idx| unsafe {
+            // We only transmute and copy during iteration, which is safe and serves
+            // as a poor man's replacement for the missing re-borrowing semantic.
+            let values: O::Value = std::mem::transmute_copy(&self.values);
```

Comment:
This makes me nervous. I think it would be better to avoid the transmute here, and push it into the `Open` trait. While this is fine for our types, I think it will be a problem if anyone's `O::Value` is something other then a borrow.